### PR TITLE
Automatically bump Forge compat base image to latest release

### PR DESCRIPTION
**Pull request by yzaccc**
## Description

- To automatically bump Forge compat base image to latest release, currently its a manual input. 
- To test the compat test against the last released docker image, we need to get the last released docker image tag.
- This step determines appropriate branch to fetch the latest docker image tag from based on the current branch.
Determine the IMAGE_TAG based on the latest release R and check for images built on branch R - 1:
      - For commits on the `aptos-release-v1.19` branch, the `IMAGE_TAG` should be the latest commit built on `aptos-release-v1.18`
      - For commits on the `main` branch, the `IMAGE_TAG` should be the latest commit on the max release `aptos-release-v<MAX>` branch
      - For commits on other branches, the IMAGE_TAG should be the latest commit on the branch's last release aptos-release-v<X.X> branch

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [x] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?

locally:
`poetry run python3 -u determine_target_branch_to_fetch_last_released_image.py main
Target branch is aptos-release-v1.20
aptos-release-v1.20`

`poetry run python3 -u determine_target_branch_to_fetch_last_released_image.py aptos-release-v1.19
Target branch is aptos-release-v1.18
aptos-release-v1.18`

Tested in current PR's CI https://github.com/aptos-labs/aptos-core/actions/runs/11007566097

Created unitests.
## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by mtrackeros — APPROVED** [review]

File: .github/workflows/docker-build-test.yaml
Comment (perryjrandall) [inline .github/workflows/docker-build-test.yaml:327]: you probably wanna quote "$(git ...)" because any spaces there will break your output

File: .github/workflows/docker-build-test.yaml
Comment (perryjrandall) [inline .github/workflows/docker-build-test.yaml:384]: A nice way to display this as well is adding it to the github step summary, then it will show up on the summary page as a note while the job is running

```
echo TARGET ... >> $GITHUB_STEP_SUMMARY
```

This can also be done from typescript / anywhere in your actions / forge etc

**Review by perryjrandall — CHANGES_REQUESTED** [review]
Main points rl;dr to unblock review

1) Please use either python / typescript (preferably typescript action) for this logic
2) Handle failure cases in the logic / add some test
3) Use last previous release instead of fall back to main when merging on main
4) Update framework image as well
5) Test using the dependent PR and keeping pull_request_target
4)

File: mempool/src/shared_mempool/tasks.rs
Comment (perryjrandall) [inline mempool/src/shared_mempool/tasks.rs:467]: because you're adding this rust change, I believe this is triggering additional tests on each PR run

You might wanna leave this out for now so it doesnt trigger the rust tests for every PR

File: .github/workflows/docker-build-test.yaml
Comment (rustielin) [inline .github/workflows/docker-build-test.yaml:317]: This header is necessary for the step that posts comments to the PR. We could derive it from the test suite though. It just needs to be unique per-PR-per-workflow, and we likely don't run the same test suite multiple times in a workflow: https://github.com/marocchino/sticky-pull-request-comment#keep-more-than-one-comment

File: .github/workflows/docker-build-test.yaml
Comment (rustielin) [reply .github/workflows/docker-build-test.yaml:357]: @perryjrandall I think your case is still a "merge to main" case though. Since commit `c` was presumably created from a PR targeting `main`. 

Let's consider another structure to be a bit more clear

```
a
|    b1 -> aptos-release-banana (tag)
b_/
|    d2 (PR)
c   |
|    d1 -> d-personal-branch (branch)
d_/
|    e1 -> aptos-release-egret (tag)
e_/
```

`d1` is the tip of a random branch `d-personal-branch` that's branched off of `main` at some point, and we want to merge `d2` in a PR into `d-personal-branch` (on top of `d1`). This PR does not target `main` branch -- it targets `d-personal-branch`. In this case we should test the PR with `aptos-release-egret`, which is the latest release in the PR source branch's history

So basically for both PRs that target non-main/non-release branches and PRs that target main, they both test against the same thing: the latest release in the target branch's git history


We should probably do something like this:
* Try to find the latest release in the PR target branch's git history
* If we can't find it in like the last ~100s commits, we may want to emit an error (probably prompt the user to rebase since it's likely they forgot to do this). This assumes that there are no large working branches of that size.
Lmk if this does/doesn't make sense, happy to sync separately

**Review by rustielin — CHANGES_REQUESTED** [review]

File: mempool/src/shared_mempool/tasks.rs
Comment (yzaccc) [reply mempool/src/shared_mempool/tasks.rs]: reverted

**Comment by yzaccc** [comment]
current job `rust-images-consensus-only-perf-test` always fail with reasons not related to this PR, so temporarily removed the compat test's dependency on this job. Need to add the dependency back before merging.

File: testsuite/determine_target_branch_to_fetch_last_released_image_test.py
Comment (perryjrandall) [inline testsuite/determine_target_branch_to_fetch_last_released_image_test.py:15]: nice testing!

File: .github/workflows/docker-build-test.yaml
Comment (perryjrandall) [inline .github/workflows/docker-build-test.yaml:313]: nit, format your comments like the other files

File: .github/actions/determine-or-use-target-branch-and-get-last-released-image/action.yaml
Comment (perryjrandall) [inline .github/actions/determine-or-use-target-branch-and-get-last-released-image/action.yaml:55]: Interesting here we really see the shortcoming of using python for github actions

Currently we do not have a good model for github action distribution and versioning (using python actions)

In typescript, we actually can build the action and tag and release them

We could in theory build a python action distribution mechanism whereby we build a binary and then in the preamble for the action download and unpack that binary and run from the composite action

Then you dont need any of this double checking out stuff / python setup

However, that is a problem for future us :)

This is fine as is, just wanna share my thoughts on this issue

**Review by perryjrandall — APPROVED** [review]
Looking good, thorough testing and clear logic <3

LFG

File: .github/workflows/docker-build-test.yaml
Comment (rustielin) [inline .github/workflows/docker-build-test.yaml:366]: so clean 😎

**Review by rustielin — APPROVED** [review]
LGTM!

**Pull request merged**
